README.md
# Zoomable Photo Gallery

A Flutter package, It would help you make has zoomable photo gallery easy.

## Getting Started

To start using this package, add `zoomable_photo_gallery` dependency to your `pubspec.yaml`

```yaml
dependencies:
  zoomable_photo_gallery: "<latest_release>"
```

## Documentation

### ZoomablePhotoGallery
This is so simple to use, it needs only images widget.
you can use asset file(jpg,png) or network image or CachedNetworkImage what ever you want.


```dart
ZoomablePhotoGallery(
    imageList: List.generate(
        imageFile.length,
        (index) => imageFile[index],
    ),
),

--- or ---

ZoomablePhotoGallery(
    controller: controller,
    initIndex: 3,
    backColor: Colors.white,
    maxZoom: 5,
    minZoom: 0.5,
    location: IndicatorLocation.BOTTOM_CENTER,
    changePage: (int index) {
        setState(() {
        currentPageIndex = index;
        });
    },
    indicator: List.generate(
        imageUrl.length,
        (index) => Container(
        margin: const EdgeInsets.symmetric(horizontal: 3),
        width: 10,
        height: 10,
        decoration: BoxDecoration(
            color: index == currentPageIndex ? Colors.red : Colors.grey,
        ),
        ),
    ).toList(),
    imageList: List.generate(
        imageUrl.length,
        (index) => CachedNetworkImage(
                imageUrl: imageUrl[index],
                fadeInDuration: Duration.zero,
                fadeOutDuration: Duration.zero,
                progressIndicatorBuilder: (_, v, p) {
                return const Center(child: CircularProgressIndicator());
                },
            )
        ),
    ),
    onTap: () {
        Navigator.pop(context);
    },
```


| Parameters | Value                                | Required | Docs                                                                                 |
| ---------- | ------------------------------------ | :------: | ------------------------------------------------------------------------------------ |
| `imageList`   | `List<Widget>`                    |   YES    | Required to image list widget                                                        |
| `initIndex`   | `int`                             |    No    | First view location of when would mounted this widget                                |
| `location`    | `IndicatorLocation`               |    No    | indicator location enum option (ex : BOTTOM_CENTER , TOP_CENTER ...)                 |
| `controller`  | `ZoomablePhotoController`         |    No    | You can controller paging in yourside. so you can make custom indicator              |
| `backColor`   | `Color`                           |    No    | You can set the exterior color.                                                      |
| `height`      | `double`                          |    No    | It is height of photo gallery , default is display of device size                    |
| `maxZoom`     | `double`                          |    No    | This is the maximum zoom possible. default is 2.5                                    |
| `minZoom`     | `double`                          |    No    | This is the minimum zoom possible. default is 0.5                                    |
| `changePage`  | `Function(int)`                   |    No    | You can received current page when page changed                                      |
| `indicator`   | `List<Widget>`                    |    No    | You can make custom indicator widget                                                 |
| `onTap`       | `Function()`                      |    No    |  You can recived image tap event                                                     |



## IndicatorLocation
indicator location enum option

```dart
enum IndicatorLocation{
    BOTTOM_CENTER,
    BOTTOM_LEFT,
    BOTTOM_RIGHT,
    TOP_CENTER,
    TOP_LEFT,
    TOP_RIGHT,
}
```
## Screens

![Sample 1](https://user-images.githubusercontent.com/36467891/179131741-6f06bb9f-ce0e-49ac-9822-3f755e6c9f0f.gif)

![Sample 2](https://user-images.githubusercontent.com/36467891/179131849-bc30a591-1587-43d1-bbc8-c50209adc905.gif)

CHANGELOG.md
## 1.0.2
* Homepage url change

## 1.0.1
* Add onTap option in ZoomablePhotoGallery.

## 1.0.0
* initial release


example/lib/main.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:zoomable_photo_gallery/zoomable_photo_gallery_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Zoomable Photo Gallery')),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(
              onPressed: () {
                Navigator.push(context,
                    MaterialPageRoute(builder: (context) => SimpleGallery()));
              },
              child: Text('Simple Option Gallery'),
            ),
            ElevatedButton(
                onPressed: () {
                  Navigator.push(context,
                      MaterialPageRoute(builder: (context) => Gallery()));
                },
                child: Text('Gallery')),
          ],
        ),
      ),
    );
  }
}

class SimpleGallery extends StatefulWidget {
  SimpleGallery({Key? key}) : super(key: key);

  @override
  State<SimpleGallery> createState() => _SimpleGalleryState();
}

class _SimpleGalleryState extends State<SimpleGallery> {
  List<Widget> imageFile = [
    Image.asset('assets/images/1.jpg'),
    Image.asset('assets/images/2.jpg'),
    Image.asset('assets/images/3.jpg'),
    Image.asset('assets/images/4.jpg'),
  ];

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Zoomable Photo Gallery')),
      body: ZoomablePhotoGallery(
        imageList: List.generate(
          imageFile.length,
          (index) => imageFile[index],
        ),
      ),
    );
  }
}

class Gallery extends StatefulWidget {
  Gallery({Key? key}) : super(key: key);

  @override
  State<Gallery> createState() => _GalleryState();
}

class _GalleryState extends State<Gallery> {
  ZoomablePhotoController controller = ZoomablePhotoController();
  int currentPageIndex = 0;
  List<String> imageUrl = [
    'https://images.pexels.com/photos/1172253/pexels-photo-1172253.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500',
    'https://img.rawpixel.com/private/static/images/website/2022-05/px142077-image-kwvvvktc.jpg?w=800&dpr=1&fit=default&crop=default&q=65&vib=3&con=3&usm=15&bg=F4F4F3&ixlib=js-2.2.1&s=0cfd8aa6e66ad55612f900d479a9d0de',
    'https://media.istockphoto.com/photos/alone-on-my-way-picture-id696708198?k=20&m=696708198&s=170667a&w=0&h=HUbGtSvBo0DfNaXh1N3PNL1M0Lzg7Sqs38e4aZ5JmJE=',
    'https://st.depositphotos.com/1005844/4918/i/600/depositphotos_49180569-stock-photo-fantasy-landscape.jpg',
    'https://p.bigstockphoto.com/GeFvQkBbSLaMdpKXF1Zv_bigstock-Aerial-View-Of-Blue-Lakes-And--227291596.jpg',
  ];

  @override
  void initState() {
    super.initState();
    controller.addListener(() {
      // print(controller.page);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Zoomable Photo Gallery')),
      body: ZoomablePhotoGallery(
        controller: controller,
        initIndex: 3,
        backColor: Colors.white,
        maxZoom: 5,
        minZoom: 0.5,
        location: IndicatorLocation.BOTTOM_CENTER,
        changePage: (int index) {
          setState(() {
            currentPageIndex = index;
          });
        },
        indicator: List.generate(
          imageUrl.length,
          (index) => Container(
            margin: const EdgeInsets.symmetric(horizontal: 3),
            width: 10,
            height: 10,
            decoration: BoxDecoration(
              color: index == currentPageIndex ? Colors.red : Colors.grey,
            ),
          ),
        ).toList(),
        imageList: List.generate(
          imageUrl.length,
          (index) => CachedNetworkImage(
            imageUrl: imageUrl[index],
            fadeInDuration: Duration.zero,
            fadeOutDuration: Duration.zero,
            progressIndicatorBuilder: (_, v, p) {
              return const Center(child: CircularProgressIndicator());
            },
          ),
        ),
        onTap: () {
          Navigator.pop(context);
        },
      ),
    );
  }
}


lib/indecator_widget.dart
import 'package:flutter/material.dart';

class IndecatorWidget extends StatelessWidget {
  final int length;
  final int activeIndex;
  final Color? activeColor;
  final Color? normalColor;
  final double? size;
  const IndecatorWidget({
    Key? key,
    required this.length,
    required this.activeIndex,
    this.activeColor,
    this.size = 5,
    this.normalColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: List.generate(
          length,
          (index) => Container(
            margin: const EdgeInsets.symmetric(horizontal: 3),
            width: size,
            height: size,
            decoration: BoxDecoration(
              borderRadius: const BorderRadius.all(Radius.circular(10)),
              color: index == activeIndex
                  ? activeColor ?? Colors.blue
                  : normalColor ?? Colors.grey,
            ),
          ),
        ).toList(),
      ),
    );
  }
}


lib/multi_gesture_detector.dart
import 'dart:math';

import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';

class MultiGestureDetector extends StatefulWidget {
  Widget child;
  Function(int touchCount, List<PointerDetail> touchDetails)? onTouchDown;
  Function(int touchCount, List<PointerDetail> touchDetails)? onTouchMove;
  Function(int touchCount, List<PointerDetail> touchDetails)? onTouchUp;
  Function()? onTap;

  MultiGestureDetector({
    Key? key,
    required this.child,
    this.onTouchDown,
    this.onTouchMove,
    this.onTouchUp,
    this.onTap,
  }) : super(key: key);

  @override
  State<MultiGestureDetector> createState() => _MultiGestureDetectorState();
}

class _MultiGestureDetectorState extends State<MultiGestureDetector> {
  Set<PointerDetail> multiPointers = {};
  bool isMultiTouchOn = false;
  @override
  Widget build(BuildContext context) {
    return Listener(
      onPointerDown: (e) {
        var x = e.position.dx.round().toDouble();
        var y = (e.position.dy).round().toDouble();
        multiPointers.add(PointerDetail(id: e.pointer, x: x, y: y));
        if (multiPointers.length > 1) {
          isMultiTouchOn = true;
        }
        if (widget.onTouchDown != null) {
          widget.onTouchDown!(multiPointers.length, multiPointers.toList());
        }
      },
      onPointerMove: (e) {
        var x = e.position.dx.round().toDouble();
        var y = (e.position.dy).round().toDouble();
        multiPointers.where((m) => m.id == e.pointer).first.updatePointer(x, y);
        if (widget.onTouchMove != null) {
          widget.onTouchMove!(multiPointers.length, multiPointers.toList());
        }
      },
      onPointerUp: (e) {
        var x = e.position.dx.round().toDouble();
        var y = (e.position.dy).round().toDouble();
        var upPoint = multiPointers.where((m) => m.id == e.pointer).first;
        upPoint.endPointer(x, y);
        var distance = upPoint.movedDistance;
        multiPointers.remove(upPoint);
        if (widget.onTouchUp != null) {
          widget.onTouchUp!(multiPointers.length, multiPointers.toList());
        }
        if (widget.onTap != null &&
            multiPointers.isEmpty &&
            !isMultiTouchOn &&
            distance < 10) {
          widget.onTap!();
        }
        if (multiPointers.isEmpty) {
          isMultiTouchOn = false;
        }
      },
      onPointerCancel: (e) {
        multiPointers.clear();
      },
      child: widget.child,
    );
  }
}

class PointerDetail extends Equatable {
  final int? id;
  Point<double>? startPoint;
  Point<double>? updatePoint;
  Point<double>? endPoint;

  PointerDetail({
    required this.id,
    required double x,
    required double y,
  })  : startPoint = Point<double>(x, y),
        updatePoint = Point<double>(x, y),
        endPoint = Point<double>(x, y);

  updatePointer(double x, double y) {
    updatePoint = Point<double>(x, y);
  }

  endPointer(double x, double y) {
    endPoint = Point<double>(x, y);
  }

  double get movedDistance => startPoint!.distanceTo(endPoint!);

  @override
  List<int> get props => [id!];
}


lib/zoomable_photo_gallery_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';
import 'package:zoomable_photo_gallery/indecator_widget.dart';
import 'package:zoomable_photo_gallery/multi_gesture_detector.dart';

class ZoomablePhotoController extends PageController {}

enum IndicatorLocation {
  BOTTOM_CENTER(Alignment.bottomCenter),
  BOTTOM_LEFT(Alignment.bottomLeft),
  BOTTOM_RIGHT(Alignment.bottomRight),
  TOP_CENTER(Alignment.topCenter),
  TOP_LEFT(Alignment.topLeft),
  TOP_RIGHT(Alignment.topRight);

  const IndicatorLocation(this.alignment);
  final Alignment alignment;
}

class ZoomablePhotoGallery extends StatefulWidget {
  final List<Widget> imageList;
  final ZoomablePhotoController? controller;
  final IndicatorLocation location;
  final int initIndex;
  final double? height;
  final Color? backColor;
  final double maxZoom;
  final double minZoom;
  final List<Widget>? indicator;
  final Function(int)? changePage;
  final Function()? onTap;

  ZoomablePhotoGallery({
    Key? key,
    required this.imageList,
    this.initIndex = 0,
    this.location = IndicatorLocation.BOTTOM_CENTER,
    this.controller,
    this.backColor = Colors.black,
    this.height,
    this.maxZoom = 2.5,
    this.minZoom = 0.5,
    this.changePage,
    this.indicator,
    this.onTap,
  }) : super(key: key);

  @override
  State<ZoomablePhotoGallery> createState() => _ZoomablePhotoGalleryState();
}

class _ZoomablePhotoGalleryState extends State<ZoomablePhotoGallery>
    with TickerProviderStateMixin {
  late int activeIndex;
  List<Widget>? fishinggramImageWidget;
  late PageController _pageController;

  final TransformationController _transformationController =
      TransformationController();
  Animation<Matrix4>? _animationReset;
  late AnimationController _controllerReset;

  @override
  void initState() {
    super.initState();
    _pageController = widget.controller ?? PageController();
    makcFishinggramImageWidget();
    _controllerReset = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 300),
    );

    activeIndex = widget.initIndex;
    if (widget.initIndex != 0) {
      WidgetsBinding.instance.addPostFrameCallback(
          (_) => _pageController.jumpToPage(widget.initIndex));
    }

    _pageController.addListener(() {
      if (_pageController.page == _pageController.page?.floor()) {
        activeIndex = _pageController.page!.toInt() % widget.imageList.length;
        if (widget.changePage != null) {
          widget.changePage!(activeIndex);
        }
        update();
      }
    });
  }

  @override
  void didUpdateWidget(covariant ZoomablePhotoGallery oldWidget) {
    makcFishinggramImageWidget();
    super.didUpdateWidget(oldWidget);
  }

  void update() => setState(() {});

  void _onAnimateReset() {
    _transformationController.value = _animationReset!.value;
    if (!_controllerReset.isAnimating) {
      _animationReset!.removeListener(_onAnimateReset);
      _animationReset = null;
      _controllerReset.reset();
    }
  }

  void _animateResetStop() {
    _controllerReset.stop();
    _animationReset?.removeListener(_onAnimateReset);
    _animationReset = null;
    _controllerReset.reset();
  }

  void _onInteractionStart(ScaleStartDetails details) {
    if (_controllerReset.status == AnimationStatus.forward) {
      _animateResetStop();
    }
  }

  void _animateResetInitialize() {
    _controllerReset.reset();
    _animationReset = Matrix4Tween(
      begin: _transformationController.value,
      end: Matrix4.identity(),
    ).animate(_controllerReset);
    _animationReset!.addListener(_onAnimateReset);
    _controllerReset.forward();
  }

  void _onInteractionEnd(ScaleEndDetails details) {
    Future.delayed(const Duration(milliseconds: 200), () {
      _animateResetInitialize();
    });
  }

  @override
  void didChangeDependencies() async {
    super.didChangeDependencies();
    await Future.delayed(const Duration(milliseconds: 200));
  }

  void makcFishinggramImageWidget() {
    fishinggramImageWidget = List.generate(
      widget.imageList.length,
      (index) => InteractiveViewer(
          boundaryMargin: const EdgeInsets.all(double.infinity),
          transformationController: _transformationController,
          minScale: widget.minZoom,
          maxScale: widget.maxZoom,
          onInteractionStart: _onInteractionStart,
          onInteractionEnd: _onInteractionEnd,
          alignPanAxis: true,
          panEnabled: false,
          child: widget.imageList[index]),
    );
  }

  @override
  void dispose() {
    _controllerReset.dispose();
    super.dispose();
  }

  bool _pagingEnabled = true;
  double startX = 0;

  void _onTouchDown(int touchCount, List<PointerDetail> touchDetails) {
    if (touchCount > 1) {
      _pagingEnabled = false;
      update();
    }
  }

  void _onTouchUp(int touchCount, List<PointerDetail> touchDetails) {
    if (touchCount == 0) {
      _pagingEnabled = true;
      update();
    }
  }

  @override
  Widget build(BuildContext context) {
    var bottom = 0.0;
    var top = 0.0;
    switch (widget.location) {
      case IndicatorLocation.BOTTOM_CENTER:
      case IndicatorLocation.BOTTOM_LEFT:
      case IndicatorLocation.BOTTOM_RIGHT:
        bottom = 20 +
            (widget.height == null ? MediaQuery.of(context).padding.bottom : 0);
        break;
      case IndicatorLocation.TOP_CENTER:
      case IndicatorLocation.TOP_LEFT:
      case IndicatorLocation.TOP_RIGHT:
        top = 20;
        break;
    }
    return MultiGestureDetector(
      onTouchDown: _onTouchDown,
      onTouchUp: _onTouchUp,
      onTap: () {
        if (widget.onTap != null) widget.onTap!();
      },
      child: Container(
        height: widget.height ?? MediaQuery.of(context).size.height,
        color: widget.backColor,
        child: Stack(
          children: [
            PageView.builder(
              physics: _pagingEnabled
                  ? const CustomPageViewScrollPhysics()
                  : const NeverScrollableScrollPhysics(),
              controller: _pageController,
              itemBuilder: (_, index) {
                return fishinggramImageWidget![index % widget.imageList.length];
              },
            ),
            const SizedBox(height: 10),
            Positioned(
              left: 20,
              right: 20,
              bottom: bottom,
              top: top,
              child: Align(
                alignment: widget.location.alignment,
                child: widget.indicator == null
                    ? IndecatorWidget(
                        size: 7,
                        length: widget.imageList.length,
                        activeIndex: activeIndex,
                        activeColor: Colors.blue,
                        normalColor: Colors.grey,
                      )
                    : Wrap(children: widget.indicator!),
              ),
            )
          ],
        ),
      ),
    );
  }
}

class CustomPageViewScrollPhysics extends ScrollPhysics {
  const CustomPageViewScrollPhysics({ScrollPhysics? parent})
      : super(parent: parent);

  @override
  CustomPageViewScrollPhysics applyTo(ScrollPhysics? ancestor) {
    return CustomPageViewScrollPhysics(parent: buildParent(ancestor)!);
  }

  @override
  SpringDescription get spring => const SpringDescription(
        mass: 50,
        stiffness: 100,
        damping: 1,
      );
}
